lib/Yield/Calls/Slices.cpp

```diff
@@ -0,0 +1,133 @@
+                                  llvm::po_end(NV{ *Entry }));
+  for (NodeView CurrentNode : llvm::reverse(PostOrder)) {
+    uint64_t &CurrentRank = Ranks[CurrentNode];
+
+    CurrentRank = 0;
+    for (auto Child : llvm::children<INV>(CurrentNode))
+      if (auto RankIterator = Ranks.find(Child); RankIterator != Ranks.end())
+        CurrentRank = std::max(RankIterator->second + 1, CurrentRank);
+  }
+
+  // For each node, select a single predecessor to keep connected to.
+  // The ranks calculated earlier are used to choose a specific one.
+  //
+  // TODO: We should consider a better selection algorithm.
+  std::unordered_map<NodeView, NodeView> RealEdges;
+  for (NodeView Node : Input.nodes()) {
+    NodeView SelectedNeighbour = nullptr;
+
+    // This selects a child with the highest rank
+    auto NodeIt = Ranks.find(Node);
+    if (NodeIt != Ranks.end()) {
+      size_t SelectedNeighbourRank = 0;
+      for (NodeView Neighbour : llvm::children<INV>(Node)) {
+        auto NeighbourIt = Ranks.find(Neighbour);
+        if (NeighbourIt != Ranks.end()) {
+
+          // Select the neighbour with the highest possible rank that is still
+          // lower than the current node's rank.
+          size_t Rank = NeighbourIt->second;
+          if (Rank < NodeIt->second && Rank >= SelectedNeighbourRank) {
+            SelectedNeighbour = Neighbour;
+            SelectedNeighbourRank = Rank;
+          }
+        }
+      }
+    }
+
+    auto [_, Success] = RealEdges.try_emplace(Node, SelectedNeighbour);
+    revng_assert(Success);
+  }
+
+  yield::Graph Result;
+  std::unordered_map<NodeView, yield::Graph::Node *> Lookup;
+
+  // Returns the version of the node from the new graph if it exists,
+  // or adds a new one to if it does not.
+  auto FindOrAddHelper = [&Result, &Lookup](NodeView OldNode) {
+    if (auto NewNode = Lookup.find(OldNode); NewNode != Lookup.end())
+      return NewNode->second;
+    else
+      return Lookup.emplace(OldNode, nodeHelper(Result, OldNode)).first->second;
+  };
+
+  // Make sure entry node is always present so that this never returns an empty
+  // graph
+  FindOrAddHelper(*Entry);
+
+  // Fill in the `Result` graph.
+  for (NodeView Node : llvm::breadth_first(NV{ *Entry })) {
+    for (auto Neighbour : llvm::children<INV>(Node)) {
+      if (Ranks.contains(Neighbour)) {
+        auto *NewNeighbour = FindOrAddHelper(Neighbour);
+        if (RealEdges.at(Node) == Neighbour) {
+          // Emit a real edge, if this is the neighbour selected earlier.
+          NewNeighbour->addSuccessor(FindOrAddHelper(Node));
+        } else {
+          // Emit a fake node otherwise.
+          auto NewNode = nodeHelper(Result, Node);
+          NewNode->NextAddress = NewNode->Address;
+          NewNeighbour->addSuccessor(NewNode);
+        }
+      }
+    }
+  }
+
+  auto EntryIterator = Lookup.find(*Entry);
+  if (EntryIterator != Lookup.end())
+    Result.setEntryNode(EntryIterator->second);
```

Comment:
Likewise, the `if` should not be needed here. The lookup should always succeed.

Reply:
I will change it into an assert.

(to explain why these things are `if`s to begin with, early on I was running into a whole lot of trouble with graph iteration and these asserts were firing all the time - so I temporarily switched them into `if`s and then forgot to go back 😅)